### ISSUE TITLE
"Database error: Error querying the database: db error: ERROR: type "citext" does not exist" but citext extension enabled on `db push`

### ISSUE THREAD

**Issue by kiily** [prisma/prisma#6944]
## Bug description

https://github.com/prisma/prisma/issues/5772 - Similar issue to what is described here.

We are trying to run the prisma commands on Vercel and we keep getting this error.

Our initial migration creates the extensions as suggested above.

```
CREATE EXTENSION IF NOT EXISTS citext;
CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
```

Weirdly, it works absolutely fine locally right now which makes it even harder to debug. The sequence of commands we are running is:

`npx prisma migrate reset --force`

`npx prisma db push --preview-feature`

To clarify, the first migration applies fine and then the second command errors. It might be worth noting that we are running against a custom postgres schema so maybe something going on in the shadow db there?

## How to reproduce

1. Reset your db with an initial migration to create the required extensions
2. Run db push to apply the rest of the schema
3. See error

## Expected behavior

`db push` after the initial migration has been applied should already know about `citext` has it has been applied before.

## Environment & setup

Currently happening on Vercel
- Database: PostgreSQL
- Node.js version: 14
- Prisma version: 2.21.2
```

```

**Comment by kiily** [comment]
Just to further this, when inspecting in PG_Admin, the main difference between the working case and the failing one is the absence of the relevant `Functions` within that schema. I wonder if the IF NOT EXISTS part of the command just looks at a single schema, establishes it exists there and does not add it anywhere else?

I also tried generating the full initial migration (i.e. create extensions + tables) and just applying that without db push and I get the same behavior; also only failing on CI

**Comment by pimeys** [comment]
I think you're right. We execute the extensions only to the shadow database, not at all to the actual one.

Plan:

- Use the describer to describe all extensions: `SELECT * FROM pg_extension`
- In the differ, find missing extensions from the actual database. Create them.

I'm still not sure would it be a problem for us when we don't have the extensions in the PSL.

**Comment by halvors** [comment]
have there been any progress on this?

**Comment by Druue** [comment]
Hey, @kiily, we have a preview feature that provides support for PostreSQL Extensions now that might help with this! You can read more in our [docs](https://www.prisma.io/docs/concepts/components/prisma-schema/postgresql-extensions).

**Comment by luisrudge** [comment]
Yeah, I'm having the same issue as @GeeWee. Works fine in the public schema, but we're creating 1 schema per integration test and then it fails with `Error: db error: ERROR: type "citext" does not exist`.

I tried create the extension before every test, but it still doesn't work:

```
await db.$executeRawUnsafe(`CREATE EXTENSION IF NOT EXISTS "citext" WITH SCHEMA "${getSchemaFromPGSQLUrl()}";`);
execSync(`npx prisma db push`, {
  env: process.env,
});
```

**Comment by luisrudge** [comment]
I found a way to make it work based on [this SO answer](https://stackoverflow.com/a/16128622/1438838), but it goes against what's recommended for production PGSQL dbs:

```
extensions = [citext(schema: "pg_catalog")]
```

It basically adds to the `pg_catalog` schema, which is magically shared between all schemas.

**Comment by janpio** [comment]
How are you creating the schemas and then migrating the schema onto these @luisrudge?

**Comment by luisrudge** [comment]
@janpio I'm not proud of it, but it works 😅 


```ts
//__mock__/db.ts
import { PrismaClient } from '@prisma/client';
export type { Prisma } from '@prisma/client';
import { URL } from 'url';
import { v4 } from 'uuid';
import { execSync } from 'child_process';
import { afterAll, beforeAll } from 'vitest';

function getPGSQLUrl() {
  if (!process.env.PGSQL_URL) {
    throw new Error('please provide a database url');
  }
  const url = new URL(process.env.PGSQL_URL);
  url.searchParams.delete('schema');
  url.searchParams.append('schema', `test-${v4()}`);
  return url.toString();
}
function getSchemaFromPGSQLUrl() {
  if (!process.env.PGSQL_URL) {
    throw new Error('please provide a database url');
  }
  const url = new URL(process.env.PGSQL_URL);
  return url.searchParams.get('schema');
}

process.env.PGSQL_URL = getPGSQLUrl();

//override the db export so it uses the dynamic PGSQL_URL
export const db = new PrismaClient();

beforeAll(() => {
  execSync(`npx prisma db push`, {
    env: process.env,
  });
});

afterAll(async () => {
  await db.$executeRawUnsafe(`DROP SCHEMA IF EXISTS "${getSchemaFromPGSQLUrl()}" CASCADE;`);
  await db.$disconnect();
});

beforeEach(async () => {
  await truncateDB();
});

async function truncateDB() {
  const tablenames = await db.$queryRaw<
    Array<{ tablename: string }>
  >`SELECT tablename FROM pg_tables WHERE schemaname=${getSchemaFromPGSQLUrl()}`;

  for (const { tablename } of tablenames) {
    if (tablename !== '_prisma_migrations') {
      try {
        await db.$executeRawUnsafe(
          `TRUNCATE TABLE "${getSchemaFromPGSQLUrl()}"."${tablename}" CASCADE;`
        );
      } catch (error) {
        console.log({ error });
      }
    }
  }
}

```

**Comment by janpio** [comment]
Interesting, that would mean that `db push` does not activate the extension as expected. Can you open a new issue with your code and also your schema you are using? Maybe we can figure this out.

**Comment by luisrudge** [comment]
@janpio I think `db push` activates the extension correctly. The problem is that you can have only a single schema for the extension and that makes it impossible to run tests in parallel. I think it's more related to this issue: https://github.com/prisma/prisma/issues/14662

**Comment by Billcountry** [comment]
To anyone who's still facing this issue. It seems the CITEXT extension doesn't work well if you have multiple schemas on your DB, it only works well for the first schema that you activate it for which caused this issue for me.

**Comment by darylteo** [comment]
Got around this issue by using a multiple databases on local docker instance. 

https://github.com/boskomijin/postgres-multiple-databases-creation

**Comment by areichman** [comment]
> To anyone who's still facing this issue. It seems the CITEXT extension doesn't work well if you have multiple schemas on your DB, it only works well for the first schema that you activate it for which caused this issue for me.

This was the issue for us as well. Somehow, our extensions were associated with a schema named `undefined`, rather than the schema for our app. We fixed our issue with the following manual steps:

- `ALTER EXTENSION citext SET SCHEMA "myapp"`
- Delete the row for the failed migration in the `_prisma_migrations` table
- Rerun the migration command
